### ISSUE TITLE
CRAN Release?

### ISSUE THREAD

**Issue by eitsupi** [pola-rs/r-polars#80]
The Rust version of Debian testing will soon be 1.65, which may allow us to build rpolars tuned to build in the R-universe.
(Note that the macOS builder's Rust version may be older than Debian's. eitsupi/prqlr#94)
https://tracker.debian.org/pkg/rustc
https://qa.debian.org/excuses.php?experimental=1&package=rustc

@sorhawell Could you potentially do a CRAN release?
Probably need to modify build scripts (like #29) and documentation (Of course I would like to contribute to that.).

**Comment by sorhawell** [comment]
Cool :)

For mac [this issue](https://github.com/pola-rs/r-polars/issues/35) will trigger a warning or error with CRAN check. I likely have to bundle some shared macOS tool or include as a requirement in DESCRIPTION.

Is it allowed on CRAN, not to release on all OS, e.g. skip mac in first round?

Regarding #29
Using CARGO_HOME is very useful to run CHECK in development without recompiling for 30 min. There is something about an environment variable called something like IS_CRAN right? I could disable CARGO_HOME for cran builds then.
I failed to rediscover the exact envvar name in docs.

**Comment by eitsupi** [comment]
> For mac [this issue](https://github.com/pola-rs/r-polars/issues/35) will trigger a warning or error with CRAN check. I likely have to bundle some shared macOS tool or include as a requirement in DESCRIPTION.

I'm sorry I don't understand the whole issue, but are you saying that due to the issue, rpolars can's update polars?
(Currently, the command check is succeeding on the main branch, right?)

> There is something about an environment variable called something like IS_CRAN right?

Yes, this must be `NOT_CRAN`. (See also extendr/rextendr#233)
Note that this is an environment variable name initiated independently by the `devtools` package, not defined by CRAN.

**Comment by sorhawell** [comment]
r-lib/rcmdchek does not support to filter errors. Many projects get the same warning on large binary size, which CRAN is not enforcing anyways.

Best advice I could find in issue track is to ignore any warnings in R CMD Check to avoid builds flagged as failed.

I found the practice sub-optimal, as I would not notice my own newly introduced warnings.
Currently I run this /check.yaml#L157) + [filter](https://github.com/pola-rs/r-polars/blob/d62d8102bf70882069e51e54f89bfce62a5ce373/inst/misc/filter_rcmdcheck.R#L12) in [workflow](https://github.com/pola-rs/r-polars/blob/d62d8102bf70882069e51e54f89bfce62a5ce373/.github/workflows) to catch all warnings and errors from R CMD Check, which were not already known.

Currently I ignore the binary size warning and the mac dependency warning/error.I  don't know how to fix it yet. However, I have never heard of any mac installation except cran Check, where it was actually a problem.

**Comment by sorhawell** [comment]
@eitsupi  Oh sorry I did not reply well to your question. 
> I'm` sorry I don't understand the whole issue, but are you saying that due to the issue, rpolars can's update polars?

We can update polars just fine. The Mac DLL warning just started at a certain commit when rust-polars introduced some new system calls. I have not solved the warning yet. I suspect that any Mac does support these system calls.




I could immediately try to submit a build to the cran windows test build system and see what we get back. I wonder if there are any limitations on build time.

**Comment by eitsupi** [comment]
Thanks for the reply.
I don't have macOS and don't have enough knowledge to help solve that problem, but I think there is a possibility that someone else can contribute.

**Comment by sorhawell** [comment]
@eitsupi 
I believe win + mac cran check warnings/fails have disappeared since around R4.3.0 I guess we should be able to make a submission to CRAN parallel to R-universe now :)

**Comment by eitsupi** [comment]
Grad to hear that!

Debian's Rust version is still 1.63, so we will not be able to do a CRAN release any time soon.

I hope to have the Debian Rust version up by summer and we can get the tasks for the CRAN release done by then.
I created a milestone https://github.com/pola-rs/r-polars/milestone/1.

**Comment by sorhawell** [comment]
Actually I think we can get [away with using  Required Rust version >=1.62](https://github.com/pola-rs/polars/pull/6680)

**Comment by sorhawell** [comment]
ohh [now I remember that namespace thing](https://github.com/pola-rs/r-polars/issues/25#issuecomment-1413764284)

**Comment by grantmcdermott** [comment]
Out of interest, I just peeked at the Debian rustc logs and 1.64 was accepted into the unstable branch earlier this week. See News here: https://packages.qa.debian.org/r/rustc.html

Last time it only took a week for 1.63 to migrate to the testing branch (which is what CRAN uses) after acceptance on unstable. So hopefully not too long now...

**Comment by eitsupi** [comment]
> Last time it only took a week for 1.63 to migrate to the testing branch (which is what CRAN uses) after acceptance on unstable. So hopefully not too long now...

Debian Rust has been completely stuck for the last six months, and it makes little sense to base the schedule on anything earlier than that.....

**Comment by eitsupi** [comment]
Migration to Rust 1.65 seems to be going well as far as the progress status is seen.
At the earliest, we expect Debian testing to be updated to Rust 1.65 in 5 days.

@sorhawell Any thoughts on submitting to CRAN?

One major issue is that the main branch assumes nightly toolchains and cannot be submitted to CRAN, so I think we need to consider whether to continue to enable simd in the main branch.

**Comment by sorhawell** [comment]
@eitsupi 
For me #262  looks to be an easier and more elegant way for us to maintain the CRAN and Runiverse release channels :) awesome !

I have not seen any benchmarks yet of +/- simd, so maybe it does not matter. Given it does give a considerable speed-boost, it would be a pity if users forget about it. I guess this can be compensated with mentioning in the README / docs, that the github binary also is a very easy way to install polars on most OS and is faster. In such case CRAN would be very important for discoverability, but users should be nudged towards the full featured polars.

**Comment by eitsupi** [comment]
@sorhawell Yeah, the #262 way with a feature seems to work.

As you say, documentation updates etc. will be needed to follow up #262. Can we merge the #262 for now?

**Comment by eitsupi** [comment]
@sorhawell @Sicheng-Pan The changes made by #233 prevent builds except for nightly-toolchain.
https://github.com/pola-rs/r-polars/actions/runs/5411224946/jobs/9833719506?pr=274

Can we revert this change for now? Or Could you fix that error?

**Comment by eitsupi** [comment]
I created a new issue about the build failure with stable Rust #276.

**Comment by eitsupi** [comment]
FYI, I sent the latest version to R-hub builder and gets the following notes.

```
* checking CRAN incoming feasibility ... [15s] NOTE
Maintainer: 'Soren Welling <sorhawell@gmail.com>'

New submission

Version contains large components (0.6.1.9000)

Possibly misspelled words in DESCRIPTION:
  Polars (2:8)
  polars (9:37)

Found the following (possibly) invalid URLs:
  URL: https://pola-rs.github.io/polars/py-polars/html/reference/expressions (moved to https://pola-rs.github.io/polars/py-polars/html/reference/expressions/)
    From: man/Expr_class.Rd
    Status: 200
    Message: OK

The Title field starts with the package name.
The Title field should be in title case. Current version is:
'Polars ported to R'
In title case that is:
'Polars Ported to R'

The Description field should not start with the package name,
  'This package' or similar.
```

**Comment by eitsupi** [comment]
Debian testing's rustc is now 1.66!

@sorhawell I am going to work on the Makevars modifications and document updates, is it possible for you to make a submission to CRAN within the next few days?
https://github.com/pola-rs/r-polars/milestone/1

**Comment by sorhawell** [comment]
> Debian testing's rustc is now 1.66!
> 
> @sorhawell I am going to work on the Makevars modifications and document updates, is it possible for you to make a submission to CRAN within the next few days? https://github.com/pola-rs/r-polars/milestone/1

Does Testing mean CRAN supports it now? Or does Testing mean in a few days 1.66 will be stable on Debian?

I can do a submission tomorrow e.g. ? :)

**Comment by eitsupi** [comment]
> Does Testing mean CRAN supports it now? Or does Testing mean in a few days 1.66 will be stable on Debian?

Debian "testing" is one of Debian's release channels for rolling releases.
The Debian used on CRAN is Debian testing.
https://cran.r-project.org/web/checks/check_flavors.html

> I can do a submission tomorrow e.g. ? :)

Wonderful!
I will send some PRs to resolve the issues added to the milestone, I hope so you can consider submitting them to CRAN once they are resolved.

**Comment by eitsupi** [comment]
After installing `cargo` and `cmake` (I didn't know this was necessary) via apt on the `r-base` Docker container (`debian:testing` based), I confirmed that latest `polars` can be installed. 🎉

**Comment by eitsupi** [comment]
I am assuming that once #304 is merged we can send the package to CRAN....

It might be better to resolve #230 before releasing, but this seemed like a pretty big change and one that would be difficult for me to fix right away.

**Comment by sorhawell** [comment]
@eitsupi  I was kidnapped into a meeting+follow  6 hours ago. What should I start on now towards cran release?

**Comment by sorhawell** [comment]
I will try that now

**Comment by sorhawell** [comment]
committed branch 0.7.0rc to check_rhub will continue tomorrow

**Comment by eitsupi** [comment]
Thanks for working that!

If my name and contact information are to be included in the DESCRIPTION, please list them as like this:
https://github.com/eitsupi/prqlr/blob/bca43ca2cd17f715f57440da7704d47c4f1ff9db/DESCRIPTION#L9

By the way, I forgot that the README needs to include the installation instructions from CRAN.
Push directly to the 0.7.0rc branch (and open the PR to track).

**Comment by eitsupi** [comment]
Opened a PR for tracking changes. #308

**Comment by sorhawell** [comment]
a todo list before I finialize submission:

- [ ] update DESCRIPTION
- [ ] merge read_x/scan_x PR
... ?

**Comment by sorhawell** [comment]
[rhub_check 1](https://builder.r-hub.io/status/polars_0.7.0.tar.gz-ebe8d6148698449c9b7d910dc061bbdb)

**Comment by eitsupi** [comment]
Oh, R-hub Windows does not have Rust installed....... (r-hub/rhub#550)
It is installed on Linux. (although it timed out the other day when I tried it)

**Comment by etiennebacher** [comment]
> Since @etiennebacher and @grantmcdermott obviously know more about submitting packages to CRAN than I do, I would appreciate it if you could tell me about good materials.

`devtools::release()` already comes with a a nice list of things to do before submitting. There's also the collaborative list I linked in #297.

**Comment by eitsupi** [comment]
As I commented on #308, I believe we are ready to submit.

If the R-hub builder could have done the check, we would not have been surprised by the results when `NOT_CRAN="true"` is set.......

**Comment by grantmcdermott** [comment]
Hey folks, sorry I haven't weighed in here yet. Public holiday here yesterday and I'm frantically wrapping up work before boarding a transatlantic flight this afternoon. I can take a proper look on the plane. But will echo Etienne's list and/or the `devtools::release` checklist if you're new to CRAN submissions. Those are good, easy to follow references... And sounds like everything might already be in order.

**Comment by eitsupi** [comment]
As noted in https://github.com/pola-rs/r-polars/pull/308#issuecomment-1641130045, CRAN release seems to be difficult at this time.
(I don't know of any easy solution to overcome the build time limitation.)

---

> This was seen using 1500% CPU on the Fedora check system during
installation.  As in
>
> * checking whether package ‘polars’ can be installed ... [6192s/525s] ERROR
>
> which has also exceeded the allowed CPU time limit on that system.
> That is a major violation of the CRAN policy, so the package will be
removed from CRAN.
>
> Another is misrepresentation of authorship -- you have ignored the policy's
> (’All components’ includes any downloaded at installation or during use.) and AFAICS not listed the authors and copyright holders of those you
download (and certainly not said who is responsible for those).

---

> I finally found this.  cargo build help says
>
>     Miscellaneous Options
>         -j N, --jobs N
>             Number of parallel jobs to run. May also be specified with the
>             [build.jobs](http://build.jobs/) config value
>             <https://doc.rust-lang.org/cargo/reference/config.html>.
> Defaults to the number of logical CPUs.
>
> So another thing for the 'Using Rust' document.

**Comment by sorhawell** [comment]
6192s seems very slow especially if it was multi-core. My older 2015 computer uses ~1200 secs on 4 physical cores.
If 525 secs is the acceptable limit, that could be a bit hard to achieve also. How do we test on their fedora machine?

Maybe then we´re back to have to ask for special permission to install from external binary.

**Comment by etiennebacher** [comment]
Hi all, sorry I wasn't available in the last few days but I'm not sure I could have helped in any case. If the CRAN release is not gonna happen soon then I agree with @eitsupi about making a github release anyway (cf #308)

> How do we test on their fedora machine?

Maybe with `rhub::check_on_fedora()`?

**Comment by etiennebacher** [comment]
Also, there's now a 10MB limit for package tarballs: https://github.com/eddelbuettel/crp/commit/6dcb0ec7064e3f73461f5538365f48d061f7fa9c

Is this another concern for us?

**Comment by sorhawell** [comment]
When submitting to cran I think a saw message stating the tarball was 500kB or so

**Comment by sorhawell** [comment]
Would it be possible to just state, that polars is not intended for the "Fedora" distibution?

**Comment by eitsupi** [comment]
As I understand it, this has nothing to do with Fedora. It simply means that the test was done on Fedora.
I would not be surprised if the same problem occurs on Debian, macOS or Windows.

**Comment by eitsupi** [comment]
> Also, there's now a 10MB limit for package tarballs: [eddelbuettel/crp@6dcb0ec](https://github.com/eddelbuettel/crp/commit/6dcb0ec7064e3f73461f5538365f48d061f7fa9c)
> 
> Is this another concern for us?

The capacity of the source would basically increase with the inclusion of images and pdf files.
The arrow package, for example, has given up including vignettes in the package to accommodate capacity limitations and keeps everything on the website only.

If we include all dependent crates in the source in the future, capacity limits could become an issue.

**Comment by david-cortes** [comment]
> As noted in [#308 (comment)](https://github.com/pola-rs/r-polars/pull/308#issuecomment-1641130045), CRAN release seems to be difficult at this time. (I don't know of any easy solution to overcome the build time limitation.)
> 
> > This was seen using 1500% CPU on the Fedora check system during
> > installation.  As in
> > 
> > * checking whether package ‘polars’ can be installed ... [6192s/525s] ERROR
> > 
> > which has also exceeded the allowed CPU time limit on that system.
> > That is a major violation of the CRAN policy, so the package will be
> > removed from CRAN.
> > Another is misrepresentation of authorship -- you have ignored the policy's
> > (’All components’ includes any downloaded at installation or during use.) and AFAICS not listed the authors and copyright holders of those you
> > download (and certainly not said who is responsible for those).
> 
> > I finally found this.  cargo build help says
> > ```
> > Miscellaneous Options
> >     -j N, --jobs N
> >         Number of parallel jobs to run. May also be specified with the
> >         [build.jobs](http://build.jobs/) config value
> >         <https://doc.rust-lang.org/cargo/reference/config.html>.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Defaults to the number of logical CPUs.
> > So another thing for the 'Using Rust' document.

Not an expert on the topic, but I _think_ the issue you are seeing is because the compilation process is using more than 1 thread (which it shouldn't), not because of the times - you can see that heavier packages with very long times builds have made it to CRAN before, such as [MLPack](https://cran.r-project.org/web/checks/check_results_mlpack.html) or [DuckDB](https://cran.r-project.org/web/checks/check_results_duckdb.html) (note that build times for e.g. the ASAN-enabled versions are much, much longer than what's shown on those pages).

Also don't know anything about cargo/rust, but downloading files during package building doesn't sound like something CRAN would like.

**Comment by sorhawell** [comment]
@david-cortes  Thank you for suggesting this interpretation. You could very well be right on that.

A list of what I think we need to address and a new submission:
 - limit cpu no.
 - use cargo vendor to bundle all ~150 crates
 - Somehow keep the resulting tarball size less than 10Mb
 - use scripts to auto-generate full authors and licence list of all 150 crates
 - publish all github dependencies as crates on crates.io or bundle that code also
 
Currently it seems CRAN is trying out new rules for packages including rust code. I know other maintainers of more smaller rust-R packages are scrambling to not get archived. I suspect CRAN has not made the last update of these rules, as the text leaves many open questions to me.

IMO CRAN focus is all on long term academic reproducibility and not on supporting new features or having a developer friendly repository. R-Universe is pretty fine to me. Currently to also get SIMD optimization, which requires rust nightly, only the github release binary will do.

I would personally prefer to not be the maintainer of the polars package on CRAN. I support anyone who would take on that role.

Who know maybe the goal is closer than expected :)

**Comment by david-cortes** [comment]
> @david-cortes Thank you for suggesting this interpretation. You could very well be right on that.
> 
> A fast list of what I think we need to address in a new submission:
> 
> * limit n cpus
> * use cargo vendor to bundle all ~150 crates
> * Somehow keep the resulting tarball size less than 10Mb
> * use scripts to auto-generate full authors and licence list of all 150 crates
> * publish all github dependencies as crates on crates.io or bundle that code also
> 
> Currently it seems CRAN is trying out new rules for packages including rust code. I know other maintainers of more smaller rust-R packages are scrambling to not get archived. I suspect CRAN has not made the last update of these rules, as the text leaves many open questions to me.
> 
> IMO CRAN focus is all on long term academic reproducibility and not on supporting new features or having a developer friendly repository. R-Universe is pretty fine to me. Currently to also get SIMD optimization, which requires rust nightly, only the github release binary will do.
> 
> I would personally prefer to not be the maintainer of the polars package on CRAN. I support anyone who would take on that role.
> 
> Who knows maybe the goal is closer than expected :)

Another potential route is to have a system dependency on `polars` as a shared library, and submit only the R wrapper to CRAN (not sure if that's how this library works though). Software bundles with complicated build dependencies tend follow that approach instead (e.g. `cbc`, `rgdal`), but it's not as easy to install the library as a user then.

**Comment by sorhawell** [comment]
@david-cortes  interesting idea. I see how that could help skipping CRAN compilation issues. I took a look at rgdal it appears to be C-code wrapping the external requirement gdal via a extern "C" ABI (I pressume).

In the case of r-polars, py-polars, it is rust code wrapping the rust-polars API and compiling into one optimized compilation unit and one binary. I have not seen yet polars as an external-binary, which could be called via some extern "C" ABI. Anyways then this project would need to be refactored a whole lot into C or have a minimal bit in rust and we would need to publish an external polars-for-R binary via some other channel.

One of aim of this package is to have no hard dependencies like `data.table`. Therefore installing package from binary takes ~15 sec and just works.

I personally wanted to learn rust and support R with some better tools for ML preprocessing to keep up with other languages. But ML in production is just not much in the scope of CRAN, I guess. I'm personally not very excited when thinking about CRAN rules. If I still were in academia and wanted to publish some classic statistical work in some R related journal, I would likely just use Rcpp and/or plain R, and think CRAN was a perfect match.

**Comment by mrwunderbar666** [comment]
Hi,

I have a vested interest in making rpolars available via CRAN. And I know the process is horrible.
I could help with preparing the package so the Linux builds will comply to their rules. I've been working on adjusting the build scripts to do most tasks automatically. (*I will show my fork soon*)

Looking at your todo list:

> limit n cpus
I've added checks in `Makevars` that infer whether it is running on CRAN and then act accordingly

> use cargo vendor to bundle all ~150 crates

There is an example for that here: https://github.com/yutannihilation/string2path
I added a package-cran command to the makefile which takes care of the vendoring process and outputs a tar that *should* be suitable for CRAN submissions. 

> Somehow keep the resulting tarball size less than 10Mb

This is impossible, and the CRAN maintainers should acknowledge that and they list it also in their [policies](https://cran.r-project.org/submit.html):

> Source package tarballs should if possible not exceed 10MB. It is much preferred that third-party source software should be included within the package (as e.g. a vendor.tar.xz file) than be downloaded at installation: if this requires a larger tarball a modestly increased limit can be requested at submission.

Currently, the resulting tar is about 34 MB.  

> use scripts to auto-generate full authors and licence list of all 150 crates

I took the script by @yutannihilation `[update_authors.R](https://github.com/yutannihilation/string2path/blob/main/update_authors.R)` which takes care of this process.

> publish all github dependencies as crates on crates.io or bundle that code also

I think this won't be necessary, because `cargo vendor` includes all files for offiline compilation

**Comment by sorhawell** [comment]
@mrwunderbar666  many thanks to take a stab at this!! <3 I'm currently on vacation and will return in a few days.

**Comment by mrwunderbar666** [comment]
After doing some research, I saw that the [hellorust](https://github.com/r-rust/hellorust) example comes with a lot of the required solutions for CRAN packaging. Therefore, I will ping you when I implemented a cleaner solution

**Comment by eitsupi** [comment]
Do you know of any packages on CRAN that download tarballs from the web?
I think that is the only problem here.

The big question is where and who should manage the tarballs anyway (I don't want to do that).

**Comment by sorhawell** [comment]
> I think this won't be necessary, because cargo vendor includes all files for offiline compilation

Oh nice I did not think of that.

>  if this requires a larger tarball a modestly increased limit can be requested at submission.

We can hope for a particle physicist reviewing if 340% is a modest increase :)

**Comment by eitsupi** [comment]
> We can hope for a particle physicist reviewing if 340% is a modest increase :)

Note that the 12MB prqlr was rejected.

**Comment by yutannihilation** [comment]
On prqlr's case, it might be worth trying to get rid of winapi, which I struggled a while ago. It's huge.

https://github.com/extendr/libR-sys/issues/107

winapi is a dependency of stacker, which is a dependency of chumsky. The dev version of stacker already switched it to windows-sys. I'm not sure how lighter windows-sys is than winapi, though.

https://github.com/rust-lang/stacker/pull/76

On polars' case, it's a crossterm's dependency, and it seems it won't be replaced very soon, unfortunately (Honestly, I'm not sure why the Rust code dedicated for an R package needs to consider Windows terminals).

https://github.com/crossterm-rs/crossterm/pull/718

**Comment by sorhawell** [comment]
Just learned about the zenodo project. It looks cool.

**Comment by eitsupi** [comment]
@yutannihilation Thank you for the detailed look at dependencies!

> On prqlr's case, it might be worth trying to get rid of winapi, which I struggled a while ago. It's huge.
> 
> [extendr/libR-sys#107](https://github.com/extendr/libR-sys/issues/107)
> 
> winapi is a dependency of stacker, which is a dependency of chumsky. The dev version of stacker already switched it to windows-sys. I'm not sure how lighter windows-sys is than winapi, though.
> 
> [rust-lang/stacker#76](https://github.com/rust-lang/stacker/pull/76)

I didn't know about this. thank you. However, I think removing this dependency is beyond my capabilities.

> On polars' case, it's a crossterm's dependency, and it seems it won't be replaced very soon, unfortunately (Honestly, I'm not sure why the Rust code dedicated for an R package needs to consider Windows terminals).
> 
> [crossterm-rs/crossterm#718](https://github.com/crossterm-rs/crossterm/pull/718)

I think r-polars has pulled in unnecessary functionality of rust-polars (it looks quite different compared to python-polars).

I think removing these unnecessary features would definitely reduce the size.
That said, I still don't think we can get around the 10MB limit.

---

I really don't want to use Zenodo. Every time we update the `Cargo.lock` file we have to upload a huge tarball to there?

**Comment by mrwunderbar666** [comment]
> I really don't want to use Zenodo. Every time we update the Cargo.lock file we have to upload a huge tarball to there?

I know it's not the most elegant solution. But I my idea is to only update the tarball with vendored source code when there is a CRAN release. This would also imply that the CRAN releases are rare. Users who want to use the latest features would use R-Universe instead

**Comment by sorhawell** [comment]
> I really don't want to use Zenodo. Every time we update the Cargo.lock file we have to upload a huge tarball to there?

I'm neutral, if it works and accepted, it seems ok. I see there is some github integration, I guess it could be incorporated in a CRAN release workflow.

>  But I my idea is to only update the tarball with vendored source code when there is a CRAN release.

I suppose it would be a new unique upload and any cran release would just be hardcoded* in the Makevars to point to the tarball it needs for installation.

* hardcoded url should be set by a release script.

> I think r-polars has pulled in unnecessary functionality of rust-polars (it looks quite different compared to python-polars).

`py-polars` and `r-polars` implementations are quite different when bridging multi-threading to the interpreter. `py-polars` can rely much more on PyO3 and python native threads, where r-polars e.g. uses crate `flume` and custom code. `r-polars` now also supports background queries which does not block the R interpreter and also supports having a pool of R interpreters to process R code in parallel. The `ipc-channel` crate is used to send R objects across processes via shared memory.

Also the dependency stack of `extendr-api` and `PyO3` is not the same.

My impression is that pulling in a handful low-level crates does not change the final binary much. It is the ~220 crates pulled in total which makes a big binary.

**Comment by eitsupi** [comment]
> I know it's not the most elegant solution. But I my idea is to only update the tarball with vendored source code when there is a CRAN release. This would also imply that the CRAN releases are rare. Users who want to use the latest features would use R-Universe instead

As far as I know, there is no way to tell when a package is built whether it is on R-universe or on CRAN.
In other words, on R-universe, builds and tests run the same as on CRAN.

**Comment by sorhawell** [comment]
> As far as I know, there is no way to tell when a package is built whether it is on R-universe or on CRAN.
 
 I have no proof of concept but since we get to choose what we submit for CRAN, we could have a github workflow something like this:
 
 - checkout r-polars reposity
 - run `prepare_for_cran.sh` (this script writes a file called e.g. `./src/build_config.yaml` and perhaps do other modifications to source code, deleteing any illegal files + uploading a tarball to a external server)
 - run some submit_to_cran() function on modified source code `./.`
 - Makevars will check if exists `./src/build_config.yaml`. If so, assign the variables as environment variables
 
 The `./src/build_config.yaml` could set envvars like "download_tarball_mode=true" and "tarball_url=https:/server/account/polars_tarball_0.8.0_0001"
 
 When not running prepare_for_cran.sh such envvars would not be set and R-universe or Github releases would proceed as normal.
 

 ... maybe I'm missing something :)

**Comment by etiennebacher** [comment]
Sorry if I missed something, but I don't understand how using Zenodo helps for CRAN submission. I understood that CRAN wants the package build to be in total isolation so that there's no download of external dependencies when the user installs the package, which is why they said that even downloading crates from crates.io was not an option.

Why is Zenodo a solution in this case? Is it considered as a "trusted source" by CRAN?

**Comment by sorhawell** [comment]
It seems from prsqlr that a tarball size of 12MB is unacceptably above a CRAN limit 10MB. We are allegedly at ~30MB? We cannot slim down to 10MB I think.

Was there line saying total isolation in new rules? I think I saw "no downloading of binaries."

**Comment by sorhawell** [comment]
> such as polars-sql. Do you think those have been completely removed?

it seems rust-polars features ipc, parquet, csv and json trigger polars-sql dependency. I don't know if we can cherry-pick sub features within rust-polars. I tried naively to replace e.g. "parquet" with the sub features ["polars-io", "polars-core/parquet", "polars-lazy/parquet", "polars-io/parquet", "polars-sql/parquet"], but the compiler said I could not do that, I think.

**Comment by mrwunderbar666** [comment]
I suggested using zenodo, after reading the [CRAN submission policies](https://cran.r-project.org/submit.html) in detail. Here are the relevant sections:

> If the sources are too large, it is acceptable to download them as part of installation, but do ensure that the download is of a fixed version rather than the latest. Only as a last resort and with the agreement of the CRAN team should a package download pre-compiled software.

Specifically for [rust it says](https://cran.r-project.org/submit.html):

> Packages wishing to use Rust code should either

> include the code in the package, or
> download a specific version from a secure and reliable site and check that the download is the expected code by some sort of checksum. The expected checksum needs to be embedded in the source package.

**Comment by etiennebacher** [comment]
> @etiennebacher Was there a line saying "in total isolation" in new rules? I think I saw "no downloading of binaries."

You're right @sorhawell, I misread the guidelines for Rust. However, as @mrwunderbar666 said above, the CRAN policies say

> Packages wishing to use Rust code should either include the code in the package, or download a specific version from a secure and reliable site 

but AFAICT nowhere in CRAN policies do they mention Zenodo. So I think before anything we should ask them directly if this would be acceptable to them (and if they agree with this submitting strategy overall).

**Comment by eitsupi** [comment]
I believe this is where Zenodo's name came up. (It is truly surprising that they equate Crates.io with GitHub.)

https://stat.ethz.ch/pipermail/r-package-devel/2023q3/009332.html

> Yutani,
>
> I'm not quite sure your reading fully matches the intent of the policy. Cargo.lock is not sufficient, it is expected that the package will provide *all* the sources, it is not expected to use cargo to resolve them from random (possibly inaccessible) places. So the package author is expected to either include the sources in the package *or* (if prohibitive due to extreme size) have a release tar ball available at a fixed, secure, reliable location (I was recommending Zenodo.org for that reason - GitHub is neither fixed nor reliable by definition).
>
> Based on that, I'm not sure I fully understand the scope of your proposal for improvement. Carlo.lock is certainly the first step that the package author should take in creating the distribution tar ball so you can fix the versions, but it is not sufficient as the next step involves collecting the related sources. We don't want R users to be involved in that can of worms (especially since the lock file itself provides no guarantees of accessibility of the components and we don't want to have to manually inspect it), the package should be ready to be used which is why it has to do that step first. Does that explain the intent better? (In general, the downloading at install time is actually a problem, because it's not uncommon to use R in environments that have no Internet access, but the download is a concession for extreme cases where the tar balls may be too big to make it part of the package, but it's yet another can of worms...).
>
> Cheers,
> Simon

**Comment by sorhawell** [comment]
@mrwunderbar666  I'm sorry to say  when updating rust-polars to 0.32.1 the minimal required version of rustc is 1.70 for without SIMD and rust nightly-2023-07-27 for with.

CRAN currently does not support 1.70, so I think we have hit another hard wall.  rust-polars have made no promise of only using the about 2 years older rustc versions released via debian as CRAN uses.

Thank you for looking into alternatives, but I think this is show stopper for now.

**Comment by eitsupi** [comment]
> rust-polars have made no promise of only using the about 2 years older rustc versions released via debian as CRAN uses.

To Debian's honor: Rust 1.66 was released in December 2022, which is still less than a year old.

**Comment by eitsupi** [comment]
> I guess for now this should be a priority for package development.

I would like to agree with that, but I think we did our best (see #308 for example) and as a result it looks impossible.

CRAN says https://github.com/pola-rs/r-polars/issues/80#issuecomment-1646735125

> This was seen using 1500% CPU on the Fedora check system during
installation.  As in
>
> * checking whether package ‘polars’ can be installed ... [6192s/525s] ERROR
>
> which has also exceeded the allowed CPU time limit on that system.

Checking in R-universe, the `polars` package does indeed have a longer build time than `arrow` and `duckdb`.
In other words, it is very likely that it cannot be built in the time allowed by CRAN.

**Comment by eitsupi** [comment]
For example, the `arrow` package turns off multiple flags in the build on CRAN, so it may be possible to reduce build time by reducing the rust-polars features in the same way here.
So someone may want to work on that.

**Comment by mrwunderbar666** [comment]
@eitsupi Sorry, mistyped! I meant it should **not** be a priority

**Comment by eitsupi** [comment]
> In other words, all existing packages on CRAN that use Rust (`gifski`, `string2path`, `hellorust`, `rshift`, `tok`) vendor dependent crates. But that is not possible here (> 10MB), and it is unclear how to work around it.

CRAN has approved 12MB of prqlr source code.
Likewise in polars, source size may no longer be an issue.

Since Debian may soon reach Rust 1.70, it may be worthwhile to try again to submit to CRAN in a few months.

My concern is that the build time is too long...
Checking on R-universe, duckdb was built in 30 minutes while polars took over 50 minutes.
https://github.com/r-universe/hannes/actions/runs/6105495163
https://github.com/r-universe/pola-rs/actions/runs/6137342310

**Comment by eitsupi** [comment]
It seems that the initial arrow package on CRAN did not include libarrow.
<https://github.com/cran/arrow/tree/7d575c8eacb1757dcb38cc4115bf0ece39690b79>

In fact, we can confirm that nothing works when this version is installed.

```r
> remotes::install_version("arrow", version = "0.14.1")
Downloading package from url: https://packagemanager.posit.co/cran/__linux__/jammy/latest/src/contrib/Archive/arrow/arrow_0.14.1.tar.gz
These packages have more recent versions available.
It is recommended to update all of them.
Which would you like to update?

1: All
2: CRAN packages only
3: None
4: withr (2.5.0 -> 2.5.1) [CRAN]


Enter one or more numbers, or an empty line to skip updates: 1
withr      (2.5.0 -> 2.5.1) [CRAN]
bit        (NA    -> 4.0.5) [CRAN]
tidyselect (NA    -> 1.2.0) [CRAN]
bit64      (NA    -> 4.0.5) [CRAN]
assertthat (NA    -> 0.2.1) [CRAN]
Installing 5 packages: withr, bit, tidyselect, bit64, assertthat
Installing packages into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
trying URL 'https://packagemanager.posit.co/cran/__linux__/jammy/latest/src/contrib/withr_2.5.1.tar.gz'
Content type 'binary/octet-stream' length 228097 bytes (222 KB)
==================================================
downloaded 222 KB

trying URL 'https://packagemanager.posit.co/cran/__linux__/jammy/latest/src/contrib/bit_4.0.5.tar.gz'
Content type 'binary/octet-stream' length 1130781 bytes (1.1 MB)
==================================================
downloaded 1.1 MB

trying URL 'https://packagemanager.posit.co/cran/__linux__/jammy/latest/src/contrib/tidyselect_1.2.0.tar.gz'
Content type 'binary/octet-stream' length 219931 bytes (214 KB)
==================================================
downloaded 214 KB

trying URL 'https://packagemanager.posit.co/cran/__linux__/jammy/latest/src/contrib/bit64_4.0.5.tar.gz'
Content type 'binary/octet-stream' length 475413 bytes (464 KB)
==================================================
downloaded 464 KB

trying URL 'https://packagemanager.posit.co/cran/__linux__/jammy/latest/src/contrib/assertthat_0.2.1.tar.gz'
Content type 'binary/octet-stream' length 52457 bytes (51 KB)
==================================================
downloaded 51 KB

* installing *binary* package ‘withr’ ...
* DONE (withr)
* installing *binary* package ‘bit’ ...
* DONE (bit)
* installing *binary* package ‘assertthat’ ...
* DONE (assertthat)
* installing *binary* package ‘tidyselect’ ...
* DONE (tidyselect)
* installing *binary* package ‘bit64’ ...
* DONE (bit64)

The downloaded source packages are in
        ‘/tmp/RtmpcEbl7u/downloaded_packages’
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
* installing *binary* package ‘arrow’ ...
* DONE (arrow)
> library(arrow)
Attaching package: ‘arrow’

The following object is masked from ‘package:utils’:

    timestamp

The following objects are masked from ‘package:base’:

    array, table

> set.seed(24)
> tab <- arrow::table(x = 1:10, y = rnorm(10))
Error in Table__from_dots(dots, schema) :
  Cannot call Table__from_dots(). Please use arrow::install_arrow() to install required runtime libraries.
```

Similarly, it may be worthwhile here to reduce functionality from the default feature to the bare minimum to reduce build time on CRAN......

**Comment by eitsupi** [comment]
I'm still considering the possibility of a CRAN release, but given the frequency of breaking changes with frequent breaking changes upstream, maybe it's better to avoid CRAN releases for a while?

**Comment by etiennebacher** [comment]
I don't see why this is a problem as long as we advertise well that there are still a lot of breaking changes upstream. The way I see it, releasing on CRAN is just a way to make it easier for people to install the package. We should emphasize that it shouldn't be used in other packages for now, but I don't see the number of breaking changes as a blocker.

If you still prefer to wait for upstream `polars` to be more stable, they plan to release 1.0 by the end of the year: https://github.com/pola-rs/polars/issues/6616#issuecomment-1736901318

**Comment by eitsupi** [comment]
> I don't see why this is a problem as long as we advertise well that there are still a lot of breaking changes upstream.

My intention was that it would take a lot of effort to fix with reverse dependency checking.

**Comment by eitsupi** [comment]
I think almost all the issues have been resolved, so if Rust on CRAN catches up to MSRV of polars, I would like to submit to CRAN again.

However, Rust on CRAN has been stagnant for more than half a year due to the outdated Fedora server becoming a bottleneck, so we don't know when that will happen.
<https://stat.ethz.ch/pipermail/r-package-devel/2023q4/010147.html>

**Comment by etiennebacher** [comment]
Supposing that the Rust version of CRAN matches the MSRV of polars, would we actually be ready to submit? If I remember correctly one of the main issues was also the package size, do we have a clear strategy for this?

**Comment by eitsupi** [comment]
I think the package size issue has now been resolved.
The size of prqlr 0.7.0 is currently 15MB, making it the third largest package on CRAN, but it was released quickly.

**Comment by etiennebacher** [comment]
Wasn't `polars` much larger than that? [This comment](https://github.com/pola-rs/r-polars/issues/80#issuecomment-1683704067) mentioned 34MB

**Comment by eitsupi** [comment]
> Wasn't `polars` much larger than that? [This comment](https://github.com/pola-rs/r-polars/issues/80#issuecomment-1683704067) mentioned 34MB

The point here is that packages larger than 5MB are usually not allowed. I don't think there is any difference between 15MB and 34MB in the sense that it exceeds 5MB. (Maybe there is, but we won't know until we ask CRAN)

**Comment by david-cortes** [comment]
> > Wasn't `polars` much larger than that? [This comment](https://github.com/pola-rs/r-polars/issues/80#issuecomment-1683704067) mentioned 34MB
> 
> The point here is that packages larger than 5MB are usually not allowed. I don't think there is any difference between 15MB and 34MB in the sense that it exceeds 5MB. (Maybe there is, but we won't know until we ask CRAN)

Not sure if this is applicable to rust packages, but if e.g. header files are required from other libraries and they make up substantial space, one potential workaround could be to create multiple packages containing only those headers under `inst`, and then add those packages as dependencies under `LinkingTo` in `DESCRIPTION`.

**Comment by eitsupi** [comment]
Thank you for your comment. Cargo supports vendoring very well and the CRAN team is aware of the package size issue, so it's probably OK for Rust to just include all the source code in the package.

**Comment by eitsupi** [comment]
Given the low update frequency of Rust on CRAN (they are currently using Fedora 36 from EOL for almost a year now, and they are at Rust 1.69 with no plans to be updated in the future) and the high update frequency of MSRV on (Rust) Polars, I don't know if we should submit polars to CRAN even if they temporarily meet the MSRV requirements.

One option would be to start CRAN releases after the infrastructure on CRAN has improved and there is a guarantee that Rust will be regularly updated.
(See <https://github.com/RConsortium/r-repositories-wg>)

**Comment by etiennebacher** [comment]
> (See [RConsortium/r-repositories-wg](https://github.com/RConsortium/r-repositories-wg))

I don't see any discussion on Rust on this repo, am I missing something or do you mean that we should initiate the discussion there?

**Comment by eitsupi** [comment]
@etiennebacher This is a problem with the current CRAN infrastructure, not with Rust support. In other words, the current CRAN may be stuck with the Rust version 1.69 forever because it does not know if Fedora 36 will be used until a week, a year, or 10 years from now.
I suspect this situation will continue until the current infrastructure, which volunteers have spent a great deal of effort to maintain, is completely renewed.

**Comment by etiennebacher** [comment]
I think we should close this issue. The main blocker is the Rust version on the current CRAN infrastructure and there's nothing we can do about this. We can reopen when there's some news about this. @eitsupi do you agree?

**Comment by eitsupi** [comment]
The Fedora machine on CRAN was recently replaced by 40, and it appears that Rust version updates have resumed.
This indicates that CRAN may be able to catch up with MSRV on polars as before, but I suspect that Rust on CRAN may stagnate again.

I have an automated Rust version checker on CRAN here, so anyone interested can subscribe to the release.
https://github.com/eitsupi/cran-rust-version
